lib/msf/ui/console/command_dispatcher/core.rb

```diff
@@ -1794,6 +1823,162 @@ def cmd_sessions(*args)
+    matching_sessions = {}
+    terms = search_term.split
+    id_searches, type_searches, checkin_searches, searches = [], [], [], []
+
+    # Group search terms by what's being searched for
+    terms.each do |term|
+      case term.split(":").first
+      when "session_id"
+        id_searches << term
+      when "session_type"
+        type_searches << term
+      when "last_checkin"
+        checkin_searches << term
+      else
+        print_error("Please provide valid search term. Given: #{term.split(":").first}")
+        return
+      end
+    end
+
+    # Make sure user is properly querying checkin
+    unless checkin_searches.length() < 2
+      unless validate_checkin_searches(checkin_searches)
+        return
+      end
+    end
+
+    # Group results by search term - OR filters
+    [id_searches, type_searches].each do |terms|
+      unless terms.empty?
+        id_matches = {}
+        terms.each do |term|
+          matches = filter_sessions_by_search(term)
+          return unless matches
+          id_matches = id_matches.merge(matches)
+        end
+        searches << id_matches
+      end
+    end
+
+    # Retrieve checkin search results. AND filter with a max length of 2
+    unless checkin_searches.empty?
+      checkin_matches = {}
+      checkin_matches = filter_sessions_by_search(checkin_searches.first)
+      if checkin_searches[1]
+        matches = filter_sessions_by_search(checkin_searches[1])
+        checkin_matches = checkin_matches.select{ |session_id, session| matches[session_id] == session }
+      end
+      searches << checkin_matches
+    end
+
+    # AND all the results together for final session list
+    unless searches.empty?
+      matching_sessions = searches.first
+      searches[1..].each do |result_set|
+        matching_sessions = matching_sessions.select{| session_id, session| result_set[session_id] == session}
+      end
+    else
+      print_error("Please provide a valid search query.")
+      return
+    end
+    matching_sessions
+  end
+
+  def validate_checkin_searches(checkin_searches)
+    if checkin_searches.length > 2
+      print_error("Too many checkin searches. Max: 2. Given: #{checkin_searches.length}")
+    end
+    valid_operators = ["before", "after"]
+    field1, operator1, value1 = checkin_searches[0].split(":")
+    field2, operator2, value2 = checkin_searches[1].split(":")
+    unless valid_operators.include?(operator1) && valid_operators.include?(operator2)
+      print_error("last_checkin can only be searched for using before or after. Ex: last_checkin:before:1m30s")
+      return false
+    end
+    if operator1 == operator2
+      print_error("Cannot search for last_checkin with two #{operator1} arguments.")
+      return false
+    end
+    if (operator1 == "before" && operator2 == "after" && parse_duration(value2) < parse_duration(value1)) || (operator1 == "after" && operator2 == "before" && parse_duration(value1) < parse_duration(value2))
```

Comment:
I think you could tidy up this if statement
you know by this point you must have a 'before' and an 'after' so checking if operator1 is 'before' should be enough to tell you what you need to know
alternatively maybe adding them to a hash like `{ 'before' => 123, 'after' => 456}` would make it easier to deal with/reason about